pkg/pma/pma.c

```diff
@@ -0,0 +1,783 @@
+///                            memory.
+/// @param[in]     pma         PMA this mapping will belong to.
+/// @param[in,out] len         Populated with the length of the new mapping. If
+///                            path is NULL, this parameter can also be used to
+///                            supply a non-default length for an anonymous
+///                            mapping.
+/// @param[out]    fd          Populated with the file descriptor of the opened
+///                            backing file. -1 if path is NULL.
+///
+///
+/// @return 0   Successfully created a new mapping.
+/// @return -1  Failed to create a new mapping.
+static_ int
+map_file_(const char *path,
+          void       *base,
+          bool        grows_down,
+          pma_t      *pma,
+          size_t     *len,
+          int        *fd);
+
+/// Get the status of the page surrounding an address. To set rather than get,
+/// see set_page_status_().
+///
+/// @param[in] addr  Address within the page in question. Must be within the
+///                  bounds of the PMA.
+/// @param[in] pma   PMA the page in question belongs to.
+static_ inline_ page_status_t
+page_status_(const void *addr, const pma_t *pma)
+{
+    size_t  pg_idx    = addr_to_page_idx_(addr, pma);
+    size_t  entry_idx = pg_idx / kPagesPerEntry;
+    size_t  bit_idx   = (pg_idx % kPagesPerEntry) * kBitsPerPage;
+    uint8_t status    = (pma->pg_status[entry_idx] >> bit_idx) & PS_MASK;
+    return status;
+}
+
+/// Set the status of the page surrounding an address. To get rather than set,
+/// see page_status_().
+///
+/// @param[in] addr    Address within the page in question. Must be within the
+///                    bounds of the PMA.
+/// @param[in] status  New status of the page in question. Must not be PS_MASK.
+/// @param[in] pma     PMA the page in question belongs to.
+static_ inline_ void
+set_page_status_(const void *addr, page_status_t status, const pma_t *pma)
+{
+    size_t  pg_idx    = addr_to_page_idx_(addr, pma);
+    size_t  entry_idx = pg_idx / kPagesPerEntry;
+    size_t  bit_idx   = (pg_idx % kPagesPerEntry) * kBitsPerPage;
+    uint8_t entry     = pma->pg_status[entry_idx];
+    pma->pg_status[entry_idx]
+        = (entry & ~(PS_MASK << bit_idx)) | (status << bit_idx);
+}
+
+/// Set the status of the page range starting at an address. To set the page
+/// status of a single page, see set_page_status_().
+///
+/// @param[in] addr    Address within the first page in question. Must be within
+///                    the bounds of the PMA.
+/// @param[in] pg_cnt  Number of pages in the range. The entirety of the page
+///                    range must be within the bounds of the PMA.
+/// @param[in] status  New status of the page range in question. Must not be
+///                    PS_MASK.
+/// @param[in] pma     PMA the page range in question belongs to.
+static_ inline_ void
+set_page_status_range_(const void   *addr,
+                       size_t        pg_cnt,
```

Comment:
`size_t` is a better choice for (unsurprisingly) sizes because it's guaranteed to be large enough to represent the largest object on the given platform. We don't want to artificially constrain the functionality of this function by restricting the number of pages it modify via an explicitly sized `uint` type. See:
- https://stackoverflow.com/questions/131803/unsigned-int-vs-size-t
- https://www.embedded.com/why-size_t-matters/